src/internal/cache/redisclient.go

```diff
@@ -0,0 +1,168 @@
+
+	DialConnectionTimeout time.Duration
+	DialReadTimeout       time.Duration
+	DialWriteTimeout      time.Duration
+}
+
+// GetRedisPool get a named redis pool
+// supported rawurl
+// redis://user:pass@redis_host:port/db
+// redis+sentinel://user:pass@redis_sentinel1:port1,redis_sentinel2:port2/monitor_name/db?idle_timeout_seconds=100
+func GetRedisPool(name string, rawurl string, param *RedisPoolParam) (*redis.Pool, error) {
+	if p, ok := knownPool.Load(name); ok {
+		return p.(*redis.Pool), nil
+	}
+	m.Lock()
+	defer m.Unlock()
+	// load again in case multi threads
+	if p, ok := knownPool.Load(name); ok {
+		return p.(*redis.Pool), nil
+	}
+
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		return nil, fmt.Errorf("bad redis url: %s, %s, %s", name, rawurl, err)
+	}
+
+	if param == nil {
+		param = &RedisPoolParam{
+			PoolMaxIdle:           0,
+			PoolMaxActive:         1,
+			PoolIdleTimeout:       time.Minute,
+			DialConnectionTimeout: time.Second,
+			DialReadTimeout:       time.Second,
+			DialWriteTimeout:      time.Second,
+		}
+	}
+	if t := u.Query().Get("idle_timeout_seconds"); t != "" {
+		if tt, e := strconv.Atoi(t); e == nil {
+			param.PoolIdleTimeout = time.Second * time.Duration(tt)
+		}
+	}
+
+	fmt.Println("get redis pool:", name, rawurl)
+	if u.Scheme == "redis" {
+		pool := &redis.Pool{
+			Dial: func() (redis.Conn, error) {
+				return redis.DialURL(rawurl)
+			},
+			TestOnBorrow: func(c redis.Conn, t time.Time) error {
+				_, err := c.Do("PING")
+				return err
+			},
+			MaxIdle:     param.PoolMaxIdle,
+			MaxActive:   param.PoolMaxActive,
+			IdleTimeout: param.PoolIdleTimeout,
+			Wait:        true,
+		}
+		return pool, nil
+	} else if u.Scheme == "redis+sentinel" {
+		return getSentinelPool(u, param, err, name)
+	} else if strings.HasPrefix(u.Scheme, "rediss") {
+		return nil, fmt.Errorf("bad redis url: not support secure redis")
+	} else {
+		return nil, fmt.Errorf("bad redis url: not support scheme %s", u.Scheme)
+	}
+}
+
+func getSentinelPool(u *url.URL, param *RedisPoolParam, err error, name string) (*redis.Pool, error) {
+	ps := strings.Split(u.Path, "/")
+	if len(ps) < 2 {
+		return nil, fmt.Errorf("bad redis sentinel url: no master name, %s %s", name, u)
+	}
+
+	fmt.Println("getSentinelPool:", u)
+	var sentinelOptions []redis.DialOption
+	if param.DialConnectionTimeout > 0 {
+		fmt.Println(name, "sentinel DialConnectionTimeout:", param.DialConnectionTimeout)
+		sentinelOptions = append(sentinelOptions, redis.DialConnectTimeout(param.DialConnectionTimeout))
+	}
+	if param.DialReadTimeout > 0 {
+		fmt.Println(name, "sentinel DialReadTimeout:", param.DialReadTimeout)
+		sentinelOptions = append(sentinelOptions, redis.DialReadTimeout(param.DialReadTimeout))
+	}
+	if param.DialWriteTimeout > 0 {
+		fmt.Println(name, "sentinel DialWriteTimeout:", param.DialWriteTimeout)
+		sentinelOptions = append(sentinelOptions, redis.DialWriteTimeout(param.DialWriteTimeout))
+	}
+
+	redisOptions := sentinelOptions
```

Comment:
Not needed here, please change sentinelOptions to redisOptions when defining the var.

Reply:
We do need two separate options.